chia/wallet/dao_wallet/dao_utils.py

```diff
@@ -0,0 +1,763 @@
+        proposal_validator,
+        proposal_timelock,
+        soft_close_length,
+        attendance_required,
+        pass_percentage,
+        self_destruct_length,
+        oracle_spend_delay,
+    ) = curried_args
+    curried_args = uncurry_proposal_validator(proposal_validator)
+    (
+        SINGLETON_STRUCT,
+        PROPOSAL_SELF_HASH,
+        PROPOSAL_MINIMUM_AMOUNT,
+        PAYOUT_PUZHASH,
+    ) = curried_args.as_iter()
+    return DAORules(
+        uint64(int_from_bytes(proposal_timelock.as_atom())),
+        uint64(int_from_bytes(soft_close_length.as_atom())),
+        uint64(int_from_bytes(attendance_required.as_atom())),
+        uint64(int_from_bytes(pass_percentage.as_atom())),
+        uint64(int_from_bytes(self_destruct_length.as_atom())),
+        uint64(int_from_bytes(oracle_spend_delay.as_atom())),
+        uint64(PROPOSAL_MINIMUM_AMOUNT.as_int()),
+    )
+
+
+# This takes the treasury puzzle and treasury solution, not the full puzzle and full solution
+# This also returns the treasury puzzle and not the full puzzle
+def get_new_puzzle_from_treasury_solution(puzzle_reveal: Program, solution: Program) -> Optional[Program]:
+    # curried_args = uncurry_treasury(puzzle_reveal)
+    # (
+    #     DAO_TREASURY_MOD_HASH,
+    #     DAO_PROPOSAL_VALIDATOR_MOD,
+    #     proposal_timelock,
+    #     soft_close_length,
+    #     attendance_required_percentage,
+    #     proposal_pass_percentage,
+    #     proposal_self_destruct_length,
+    #     oracle_spend_delay,
+    # ) = curried_args
+    if solution.rest().rest().first() != Program.to(0):
+        # Proposal Spend
+        mod, curried_args = solution.at("rrf").uncurry()
+        if mod == DAO_UPDATE_PROPOSAL_MOD:
+            (
+                DAO_TREASURY_MOD_HASH,
+                DAO_PROPOSAL_VALIDATOR,
+                proposal_timelock,
+                soft_close_length,
+                attendance_required,
+                pass_percentage,
+                self_destruct_length,
+                oracle_spend_delay,
+            ) = curried_args.as_iter()
+            return DAO_TREASURY_MOD.curry(
+                DAO_TREASURY_MOD_HASH,
+                DAO_PROPOSAL_VALIDATOR,
+                proposal_timelock,
+                soft_close_length,
+                attendance_required,
+                pass_percentage,
+                self_destruct_length,
+                oracle_spend_delay,
+            )
+        else:
+            return puzzle_reveal
+    else:
+        # Oracle Spend - treasury is unchanged
+        return puzzle_reveal
+
+
+# This takes the proposal puzzle and proposal solution, not the full puzzle and full solution
+# This also returns the proposal puzzle and not the full puzzle
+def get_new_puzzle_from_proposal_solution(puzzle_reveal: Program, solution: Program) -> Optional[Program]:
+    # Check if soft_close_length is in solution. If not, then add votes, otherwise close proposal
+    if len(solution.as_python()) == 1:
```

Comment:
`.as_python` will be deprecated as can cause DOS attack, can we use something else here?